packages/driver/src/cli.mts

```diff
@@ -0,0 +1,290 @@
+#!/usr/bin/env node
+import { execSync, type ExecSyncOptions } from "node:child_process";
+import { createWriteStream } from "node:fs";
+import * as os from "node:os";
+import * as fs from "node:fs/promises";
+import * as path from "node:path";
+import * as process from "node:process";
+import * as semver from "semver";
+import envPaths from "env-paths";
+import Debug from "debug";
+import which from "which";
+
+const debug = Debug("edgedb:cli");
+
+const EDGEDB_PKG_ROOT = "https://packages.edgedb.com";
+const CACHE_DIR = envPaths("edgedb").cache;
+const TEMPORARY_CLI_PATH = path.join(CACHE_DIR, "/edgedb-cli");
+const CLI_LOCATION_CACHE_FILE_PATH = path.join(CACHE_DIR, "/cli-location");
+
+interface Package {
+  name: string;
+  version: string;
+  revision: string;
+  installref: string;
+}
+
+try {
+  await main(process.argv.slice(2));
+  process.exit(0);
+} catch (err) {
+  console.error(err);
+  if (
+    typeof err === "object" &&
+    err !== null &&
+    "code" in err &&
+    typeof err.code === "number"
+  ) {
+    process.exit(err.code);
+  }
+
+  process.exit(1);
+}
+
+async function main(args: string[]) {
+  debug("Starting main function with args:", args);
+  const cliLocation =
+    (await whichEdgeDbCli()) ??
+    (await getCliLocationFromCache()) ??
+    (await installEdgeDbCli()) ??
+    null;
+
+  if (cliLocation === null) {
+    throw Error("Failed to find or install EdgeDB CLI.");
+  }
+
+  return runEdgeDbCli(args, cliLocation);
+}
+
+async function installEdgeDbCli(): Promise<string> {
+  debug("Installing EdgeDB CLI...");
+  await downloadCliPackage();
+
+  const installDir = getInstallDir(TEMPORARY_CLI_PATH);
+  const binaryPath = path.join(installDir, "edgedb");
+  await fs.writeFile(CLI_LOCATION_CACHE_FILE_PATH, binaryPath, {
+    encoding: "utf8",
+  });
+  debug("CLI installed at:", binaryPath);
+
+  try {
+    await fs.access(binaryPath, fs.constants.F_OK);
+  } catch {
+    debug("Self-installing EdgeDB CLI...");
+    selfInstallEdgeDbCli(TEMPORARY_CLI_PATH);
+  }
+  return binaryPath;
+}
+
+async function downloadCliPackage() {
+  console.log("No EdgeDB CLI found, downloading CLI package...");
```

Comment:
This is the only actual potential output unless you run the command with `DEBUG=edgedb`.

Reply:
Should we only print this if it's running in a terminal? (https://nodejs.org/docs/latest/api/tty.html#tty) If you're running this command in a script, it might be expecting a certain output (particularly if it's using the json flag), which printing this line might break.

Reply:
Yeah that seems reasonable! Will fix.

Reply:
@jaclarke ⚡ [`e873d0f` (#931)](https://github.com/edgedb/edgedb-js/pull/931/commits/e873d0f7de828206a727dd27a377406dff386496)

Also skipped the `console.error` if not TTY. That writes to stderr, so hopefully it was safe before, but seems like a reasonable choice too. Any opinions on that?

Reply:
I would have thought that we'd still want to output the error on stderr even in scripts? The command has already failed, and we're returning a non-zero exit code, so to me it seems more useful to output what the error was.

Reply:
Yeah, that seems fair. There might be some confusion if the `edgedb` binary outputs some error text and then we also output some error text here, but I guess we already have that confusion, and otherwise, any of the code in the wrapper wouldn't output anything on errors in non-TTY environments. I'll add it back and we can iterate on this if we get feedback on it being confusing or inconvenient to have to pipe stderr to `/dev/null` or something.